### PR TITLE
BaSP-MR-28: CLASS TEST - update and delete

### PR DESCRIPTION
**Pull request by araefe1992**
1. In the classes.js seed file, added  a new class object to  be used for unit testing, that include all the necessary fields and data required for the tests.
2. Created tests for PUT method: Wrote the coverage test cases to verify the functionality of the update method for the Class model.
3. Created tests for DELETE method: Wrote the coverage test cases to verify the functionality of the delete method for the Class model.

### CODE REVIEW

File: src/tests/classes.test.js [hunk 1]
```diff
@@ -1,5 +1,51 @@
...
+  day: 'tuesday',
+};
+
+beforeAll(async () => {
+  await Class.collection.insertMany(classesSeed);
+});
+
+describe('PUT /api/classes/:id', () => {
+  test('Should respond with a  200 status, class updated', async () => {
+    const response = await request(app).put('/api/classes/64693c420b3782bf437c382f').send(mockClass);
```
Comment (nicolasblasco) [inline src/tests/classes.test.js:26]: What happens if we change the class ID in the seed file? I think we should capture the id of the first class from the seed into a variable. Then we can use it for our tests.

File: src/tests/classes.test.js [hunk 2]
```diff
@@ -1,5 +1,51 @@
...
+    expect(response.status).toBe(200);
```
Comment (nicolasblasco) [inline src/tests/classes.test.js:27]: Don't you think we can test the body of the response? We can test the message, data, and error of the body of the response. 

We should do this for all of our tests.

File: src/tests/classes.test.js [hunk 3]
```diff
@@ -1,5 +1,51 @@
...
+  });
+
+  test('Should respond with a 400 status, Id is invalid', async () => {
+    const response = await request(app).put('/api/classes/64693c420b3782bf437c382r').send(mockClass);
```
Comment (nicolasblasco) [inline src/tests/classes.test.js:31]: I think we could create a variable called `mockedInvalidId` and use it as the parameter of the route

**Review by nicolasblasco — CHANGES_REQUESTED** [review]
We should also add a description to our pull request

File: src/tests/classes.test.js [hunk 4]
```diff
@@ -1,5 +1,51 @@
...
+
+  test('Should respond with a 400 status, there is missing fields', async () => {
+    const response = await request(app).put('/api/classes/64693c420b3782bf437c382f').send(mockClassInvalid);
+    expect(response.status).toBe(400);
+  });
+});
+
+describe('DELETE /api/classes/:id', () => {
+  test('Should respond with a  200 status, class deleted', async () => {
+    const response = await request(app).delete('/api/classes/64693c420b3782bf437c382f').send(mockClass);
```
Comment (JuanManuel-Blois) [inline src/tests/classes.test.js:43]: The mockClass is not necessary in the .send(), at the same time you should put it in the URL instead of the hard coded id

File: src/tests/classes.test.js [hunk 5]
```diff
@@ -1,5 +1,51 @@
...
+    expect(response.status).toBe(200);
+  });
+
+  test('Should respond with a 400 status, invalid Id', async () => {
+    const response = await request(app).delete('/api/classes/64693c420b3782bf437c3829').send(mockClass);
```
Comment (JuanManuel-Blois) [inline src/tests/classes.test.js:48]: The mockClass is not necessary in the .send(), at the same time you should put the invalidMockClass in the URL instead of the hard coded id

**Comment by araefe1992** [comment]
I only modified the seeds and the tests files, this files are because i pulled from an old commit and they just appeared in my code:
-   src/controllers/classes.js
- src/controllers/subscriptions.js
- src/models/Class.js
- src/models/Subscription.js
- src/validations/subscriptions.js

File: src/tests/classes.test.js [hunk 6]
```diff
@@ -1,5 +1,116 @@
...
+
+describe('PUT /api/classes/:id', () => {
+  test('Should respond with a  200 status, class updated', async () => {
+    const response = await request(app).put(`/api/classes/${ClassId}`).send(mockClass);
+    expect(response.status).toBe(200);
+    expect(response.body.data).toBeDefined();
+    expect(response.body.error).toBeFalsy();
+    expect(response.body.message).toBe('Class was updated succesfully');
+  });
+
```
Comment (JuanManuel-Blois) [inline src/tests/classes.test.js:46]: We should have a case where the format of the id is correct but doesnt exist

**Review by JuanManuel-Blois — APPROVED** [review]

File: src/tests/classes.test.js [hunk 7]
```diff
@@ -1,5 +1,134 @@
...
+    expect(response.body.data).toEqual({});
+    expect(response.body.error).toBeFalsy();
+    expect(response.body.message).toBe('Class could not be found and updated!');
+  });
+
+  test('Should respond with a 400 status, Id is invalid', async () => {
+    const response = await request(app).put('/api/classes/5454dgd').send(mockClass);
+    expect(response.status).toBe(400);
+    expect(response.body.error).toBeTruthy();
+    expect(response.body.message).toBe('This is not a valid object Id');
```
Comment (nicolasblasco) [inline src/tests/classes.test.js:64]: We should also test the data of the body is empty. We could use to `toBeEqual({})`

File: src/tests/classes.test.js [hunk 8]
```diff
@@ -1,5 +1,134 @@
...
+  });
+
+  test('Should respond with a 400 status, Id is not found or non-existent', async () => {
+    const response = await request(app).put(`/api/classes/${nonExistentClassId}`).send(mockClass);
+    expect(response.status).toBe(400);
+    expect(response.body.error).toBeTruthy();
+    expect(response.body.message).toBe('This is not a valid object Id');
```
Comment (nicolasblasco) [inline src/tests/classes.test.js:71]: We should also test the data of the body is empty. We could use to `toBeEqual({})`

File: src/tests/classes.test.js [hunk 9]
```diff
@@ -1,5 +1,134 @@
...
+    expect(response.body.data).toBeDefined();
+    expect(response.body.error).toBeFalsy();
+    expect(response.body.message).toBe('Class deleted succesfully');
+  });
+
+  test('Should respond with a 400 status, invalid Id', async () => {
+    const response = await request(app).delete('/api/classes/6469chs').send();
+    expect(response.status).toBe(400);
+    expect(response.body.error).toBeTruthy();
+    expect(response.body.message).toBe('This is not a valid object Id');
```
Comment (nicolasblasco) [inline src/tests/classes.test.js:110]: We should also test the data of the body is empty. We could use to `toBeEqual({})`

File: src/tests/classes.test.js [hunk 10]
```diff
@@ -1,5 +1,134 @@
...
+    expect(response.body.data).toEqual({});
+    expect(response.body.error).toBeFalsy();
+    expect(response.body.message).toBe('Class could not be found and deleted!');
+  });
+
+  test('Should respond with a 400 status, Id is not found or non-existent', async () => {
+    const response = await request(app).delete(`/api/classes/${nonExistentClassId}`).send(mockClass);
+    expect(response.status).toBe(400);
+    expect(response.body.error).toBeTruthy();
+    expect(response.body.message).toBe('This is not a valid object Id');
```
Comment (nicolasblasco) [inline src/tests/classes.test.js:126]: We should also test the data of the body is empty. We could use to `toBeEqual({})`

File: src/tests/classes.test.js [hunk 11]
```diff
@@ -1,5 +1,134 @@
...
+    expect(response.body.error).toBeTruthy();
+    expect(response.body.message).toBe('There was an error: "activityId" is required');
+  });
+
+  test('Should respond with a 500 status, server error', async () => {
+    jest.spyOn(Class, 'findByIdAndUpdate').mockRejectedValue(new Error('Something went wrong'));
+    const response = await request(app).put(`/api/classes/${ClassId}`).send(mockClass);
+    expect(response.status).toBe(500);
+  });
+});
```
Comment (nicolasblasco) [inline src/tests/classes.test.js:95]: We could also test the response.body when there is a server error

File: src/tests/classes.test.js [hunk 12]
```diff
@@ -1,5 +1,134 @@
...
+    expect(response.status).toBe(400);
+    expect(response.body.error).toBeTruthy();
+    expect(response.body.message).toBe('This is not a valid object Id');
+  });
+
+  test('Should respond with a 500 status, server error', async () => {
+    jest.spyOn(Class, 'findByIdAndDelete').mockRejectedValue(new Error('Something went wrong'));
+    const response = await request(app).delete(`/api/classes/${ClassId}`).send();
+    expect(response.status).toBe(500);
+  });
```
Comment (nicolasblasco) [inline src/tests/classes.test.js:133]: We could also test the response.body when there is a server error

**Review by nicolasblasco — CHANGES_REQUESTED** [review]

**Review by nicolasblasco — APPROVED** [review]
